registry/handlers/referrers.go

```diff
@@ -0,0 +1,196 @@
+		return
+	}
+}
+
+func (h *referrersHandler) generateReferrersList(ctx context.Context, subjectDigest digest.Digest, artifactType string) ([]v1.Descriptor, error) {
+	dcontext.GetLogger(ctx).Debug("(*referrersHandler).generateReferrersList")
+	var referrers []v1.Descriptor
+	repo := h.Repository
+	manifests, err := repo.Manifests(ctx)
+	if err != nil {
+		return nil, err
+	}
+	blobStatter := h.registry.BlobStatter()
+	rootPath := storage.GetReferrersSearchPath(repo.Named().Name(), subjectDigest)
+	err = enumerateReferrerLinks(ctx,
+		rootPath,
+		h.driver,
+		blobStatter,
+		func(referrerDigest digest.Digest) error {
+			man, err := manifests.Get(ctx, referrerDigest)
+			if err != nil {
+				return err
+			}
+			artifactManifest, ok := man.(*ociartifact.DeserializedManifest)
+			if !ok {
+				// The PUT handler would guard against this situation. Skip this manifest.
+				return nil
+			}
+			extractedArtifactType := artifactManifest.ArtifactType
+			// filtering by artifact type or bypass if no artifact type specified
+			if artifactType == "" || extractedArtifactType == artifactType {
+				desc, err := blobStatter.Stat(ctx, referrerDigest)
+				if err != nil {
+					return err
+				}
+				desc.MediaType, _, _ = man.Payload()
+				artifactDesc := v1.Descriptor{
+					MediaType:    desc.MediaType,
+					Size:         desc.Size,
+					Digest:       desc.Digest,
+					ArtifactType: extractedArtifactType,
+					Annotations:  artifactManifest.Annotations,
+				}
+				referrers = append(referrers, artifactDesc)
+			}
+			return nil
+		})
+	if err != nil {
+		switch err.(type) {
+		case driver.PathNotFoundError:
+			return nil, nil
+		}
+		return nil, err
+	}
+	return referrers, nil
+}
+
+func enumerateReferrerLinks(ctx context.Context,
+	rootPath string,
+	stDriver driver.StorageDriver,
+	blobstatter distribution.BlobStatter,
+	ingestor func(digest digest.Digest) error) error {
+
+	return stDriver.Walk(ctx, rootPath, func(fileInfo driver.FileInfo) error {
+		// exit early if directory...
+		if fileInfo.IsDir() {
+			return nil
+		}
+		filePath := fileInfo.Path()
+
+		// check if it's a link
+		_, fileName := path.Split(filePath)
+		if fileName != "link" {
+			return nil
+		}
+
+		// read the digest found in link
+		digest, err := readlink(ctx, filePath, stDriver)
+		if err != nil {
+			return err
+		}
+
+		// ensure this conforms to the linkPathFns
+		_, err = blobstatter.Stat(ctx, digest)
+		if err != nil {
+			// we expect this error to occur so we move on
+			if err == distribution.ErrBlobUnknown {
+				return nil
+			}
+			return err
+		}
+
+		return ingestor(digest)
+	})
+}
+
+func readlink(ctx context.Context, path string, stDriver driver.StorageDriver) (digest.Digest, error) {
+	content, err := stDriver.GetContent(ctx, path)
+	if err != nil {
+		return "", err
+	}
+
+	linked, err := digest.Parse(string(content))
+	if err != nil {
+		return "", err
+	}
+
+	return linked, nil
```

Comment:
code pattern can be applied.
```suggestion
	return digest.Parse(string(content))
```

Reply:
changed accordingly.